### PR TITLE
Add new module manageiq_tags

### PR DESCRIPTION
**Pull request by yaacov**
**SUMMARY**

ManageIQ is an open source management platform for Hybrid IT.

This change is adding:

    manageiq_tags module, responsible for tags management in ManageIQ

All the modules, including docs, tests and usage examples can be found here

Currently, the only requirement for the modules is manageiq-api-client-python

**ISSUE TYPE**

    New Module Pull Request

**COMPONENT NAME**

manageiq_tags.py (module)

**ANSIBLE VERSION**

ansible 2.5.0.0

### CODE REVIEW

**Comment by yaacov** [comment]
@gtanzillo @abellotti (ManageIQ team), @zgalor @cben @simon3z (Containers Management team) @joelddiaz (Ops team) - please review.

Reviewers from ManageIQ: Gregg Tanzillo (@gtanzillo) and Alberto Bellotti (@abellotti).
Maintainers: @cben @zgalor @yaacov

**Comment by yaacov** [comment]
ready_for_review

**Comment by yaacov** [comment]
shipit

File: lib/ansible/modules/remote_management/manageiq/manageiq_tags.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    def __init__(self, manageiq, resource_type, resource_id):
+        self.manageiq = manageiq
+
+        self.module = self.manageiq.module
+        self.api_url = self.manageiq.api_url
+        self.client = self.manageiq.client
+
+        self.resource_type = resource_type
+        self.resource_id = resource_id
+        self.resource_url = self.calc_resource_url()
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_tags.py]: nit: This method runs before initilization is complete, but depends on some attributes already being set. Such dependencies can be hard to follow — though in this case it's straightforward.
=> I'd just inline it here; if not consider passing resource_type/id as arguments into the method.
=> Better yet, is there a way for the helper func querying by name to return resource href?

File: lib/ansible/modules/remote_management/manageiq/manageiq_tags.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                resource_type=self.resource_type,
+                error=e)
+            self.module.fail_json(msg=msg)
+
+        tags = response.get('resources', [])
+        tags_set = set([tag['name'] for tag in tags])
+        return tags_set
+
+    def post_tags(self, tags, action):
+        """ Post the tags for the resource using ManageIQ rest api
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_tags.py]: "Post the tags" is not very clear (and the fact it uses POST not interesting).  "Perform assign/unassign action" ?

More importantly I'd say a few words about what `tags` is.

File: lib/ansible/modules/remote_management/manageiq/manageiq_tags.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    resource_type = manageiq_entities()[resource_type_key]
+
+    # create the manageiq object
+    manageiq = ManageIQ(module)
+
+    # query resource id,
+    # fail if resource does not exist
+    resource_id = query_resource_id(manageiq, resource_type, resource_name)
+
+    # create the manageiq tag object
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_tags.py]: IMHO these comments are pretty 1:1 to what the code does and can be dropped (only "fail if resource does not exist" above is helpful)
this one specifically is confusing, might sound like it's actually creating a tag in MIQ, while it's just instantiating our class...

**Review by cben — COMMENTED** [review]
Good overall, minor suggestions

File: lib/ansible/modules/remote_management/manageiq/manageiq_tags.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        url = '{resource_url}/tags'.format(resource_url=self.resource_url)
+        try:
+            response = self.client.post(url, action=action, resources=tags)
+        except Exception as e:
+            msg = "Failed to {action} tag: {error}".format(
+                action=action,
+                error=e)
+            self.module.fail_json(msg=msg)
+
+        # check all entities in resoult to be successfull
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_tags.py]: ~~typo result~~ fixed

**Comment by cben** [comment]
Very nice :+1:

shipit

**Review by cben — APPROVED** [review]

**Review by zgalor — APPROVED** [review]
LGTM
shipit

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
bot_status

**Comment by joelddiaz** [comment]
@yaacov it is sometimes useful to be able to list the current state of an object. For example, the ec2_tags module (https://docs.ansible.com/ansible/latest/ec2_tag_module.html ) allows absent, present, and list. 'list' will simply query and return all tags assigned to an object.

Otherwise, the module looks good. It worked well for me when adding/removing tags to a provider.

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
@cben @zgalor we need new shipits :-)

**Comment by yaacov** [comment]
shipit

**Comment by yaacov** [comment]
shipit

File: lib/ansible/modules/remote_management/manageiq/manageiq_tags.py [hunk 5]
```diff
@@ -148,21 +148,43 @@ def full_tag_name(self, tag):
...
+        categorization = tag.get('categorization', {})
+
+        return dict(
+            full_name=full_name,
+            name=categorization.get('name'),
+            display_name=categorization.get('display_name'),
+            category=categorization.get('category', {}).get('name'))
+
     def query_resource_tags(self):
         """ Returns a set of the full tag names assigned to the resource
```
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_tags.py]: this docstring now wrong

**Comment by yaacov** [comment]
shipit

**Comment by cben** [comment]
LGTM :+1:
shipit

@joelddiaz The new `state: list` mode returns list of dictionaries like this:
`{"full_name": "/managed/cat/entry", category: "cat", name: "entry", display_name: "Category : Entry"}`
It's superset of what you can feed back into `tags:` (and I think you can even feed them back as-is with the extra full_name, display_name).

**Comment by yaacov** [comment]
@zgalor  :sheep: :it:

**Comment by zgalor** [comment]
shipit
:leaves:  :1st_place_medal:  :wolf: alive and the :sheep: are never safe

**Comment by joelddiaz** [comment]
results from 'state: list' run looks good to me
